spec/services/price_policies/time_based_price_calculator_spec.rb

```diff
@@ -3,196 +3,323 @@
+        # Saturday at 2am
+        let(:end_at) { Time.zone.local(2017, 4, 29, 2, 0) }
+        # 1 hour in 0%, 7 hours in 10%, 2 hours in 25%
+        it { is_expected.to eq(0.88) }
+      end
 
-      it { is_expected.to be_nil }
     end
   end
 
-  describe "#calculate_discount" do
-    subject(:discount) { calculator.calculate_discount(start_at, end_at) }
+  context "when product has duration pricing mode" do
+    let(:product) { create(:setup_instrument, pricing_mode: "Duration") }
 
-    describe "when the entire time is within a zero discount" do
-      let(:start_at) { Time.zone.local(2017, 4, 27, 12, 0) }
-      let(:end_at) { Time.zone.local(2017, 4, 27, 14, 0) }
-      it { is_expected.to eq(1) }
-    end
+    describe "#calculate" do
+      subject(:costs) { calculator.calculate(start_at, end_at) }
 
-    describe "when it overlaps into evening" do
-      let(:start_at) { Time.zone.local(2017, 4, 27, 16, 0) }
+      describe "with no duration rates set" do
+        let(:options) { { usage_rate: 120 } }
+        let(:start_at) { Time.zone.local(2017, 4, 27, 12, 0) }
+        let(:end_at) { start_at + 1.hour }
 
-      describe "one hour in zero percent, one hour in 10%" do
-        let(:end_at) { Time.zone.local(2017, 4, 27, 18, 0) }
-        it { is_expected.to eq(0.95) }
+        it "uses usage_rate as rate" do
+          is_expected.to eq(cost: 120, subsidy: 0)
+        end
       end
 
-      describe "one hour in zero percent, and three hours in 10%" do
-        let(:end_at) { Time.zone.local(2017, 4, 27, 20, 0) }
-        it { is_expected.to eq(0.925) }
-      end
-    end
+      context "with duration rates" do
+        context "one of them with minimum duration of 0 hours" do
+          context "and rounded to hour time range" do
+            let(:options) { { usage_rate: 120 } }
+            let(:start_at) { Time.zone.local(2017, 4, 27, 12, 0) }
+            let(:end_at) { start_at + 1.hour }
 
-    describe "when it overlaps multiple schedule rules" do
-      # Friday at 4pm
-      let(:start_at) { Time.zone.local(2017, 4, 28, 16, 0) }
-      # Saturday at 2am
-      let(:end_at) { Time.zone.local(2017, 4, 29, 2, 0) }
-      # 1 hour in 0%, 7 hours in 10%, 2 hours in 25%
-      it { is_expected.to eq(0.88) }
-    end
+            it "uses duration rates rate" do
+              create(:duration_rate, instrument: product, min_duration: 0, rate: 60)
 
-  end
+              is_expected.to eq(cost: 60)
+            end
+          end
+
+          context "and not rounded to hour time range" do
+            let(:options) { { usage_rate: 120 } }
+            let(:start_at) { Time.zone.local(2017, 4, 27, 12, 0) }
+            let(:end_at) { start_at + 1.hour + 15.minute }
 
+            it "uses duration rates rate" do
+              create(:duration_rate, instrument: product, min_duration: 0, rate: 60)
+
+              is_expected.to eq(cost: 60 + 15)
+            end
+          end
+        end
```

Comment:
add a case for "Reservation time must be at least 1 minute within the time range in order to get the rate."